backend/pkg/kafka/consumer.go

```diff
@@ -99,100 +114,84 @@ func (s *Service) FetchMessages(ctx context.Context, progress IListMessagesProgr
 		kgo.ConsumePartitions(partitionOffsets),
 	)
 
-	// 4. Consume messages and check if filter code accepts the given records
+	// 4. Start go routine that consumes messages from Kafka and produces these records on the jobs channel so that these
+	// can be decoded by our workers.
+	go s.consumeKafkaMessages(workerCtx, client, consumeRequest, jobs)
+
+	// 5. Receive decoded messages until our request is satisfied. Once that's the case we will cancel the context
+	// that propagate to all the launched go routines.
 	messageCount := 0
 	remainingPartitionRequests := len(consumeRequest.Partitions)
+	for msg := range resultsCh {
+		// todo: Since a 'kafka message' is likely transmitted in compressed batches this is not really accurate
+		progress.OnMessageConsumed(msg.MessageSize)
+
+		if msg.IsMessageOk {
+			messageCount++
+			progress.OnMessage(msg)
+		}
+
+		partitionReq := consumeRequest.Partitions[msg.PartitionID]
+		if msg.Offset >= partitionReq.EndOffset {
+			remainingPartitionRequests--
+		}
+
+		// Do we need more messages to satisfy the user request? Return if request is satisfied
+		isRequestSatisfied := messageCount == consumeRequest.MaxMessageCount || remainingPartitionRequests == 0
+		if isRequestSatisfied {
+			// This will stop all the launched go routines for this consume request
+			cancel()
+			return nil
+		}
+	}
+
+	return nil
+}
+
+func (s *Service) consumeKafkaMessages(ctx context.Context, client *kgo.Client, consumeReq TopicConsumeRequest, jobs chan<- *kgo.Record) {
 	for {
 		select {
 		case <-ctx.Done():
-			return fmt.Errorf("interruppted message consuming because context is cancelled")
+			return
 		default:
 			fetches := client.PollFetches(ctx)
 			errors := fetches.Errors()
 			if len(errors) > 0 {
-				return fmt.Errorf("'%v' errors while fetching records: %w", len(errors), errors[0].Err)
+				s.Logger.Error("errors while fetching records",
+					zap.String("topic_name", errors[0].Topic),
+					zap.Error(errors[0].Err))
+				continue
```

Comment:
There can be errors and _also_ be records to process, so this continue should not be here. Also, there can be more than one error, but only the first error is printed -- is that intentional?